pageserver/src/repository/inmemory.rs

```diff
@@ -0,0 +1,299 @@
+use anyhow::{bail, Context, Result};
+use bytes::{Bytes};
+use log::*;
+
+use std::collections::{HashMap};
+use std::sync::{Arc, Mutex};
+use std::sync::atomic::AtomicU64;
+use std::time::{Duration};
+
+use crate::repository::{BufferTag, RelTag};
+use crate::repository::{Repository, Timeline};
+use crate::repository::WALRecord;
+use crate::waldecoder::Oid;
+use crate::walredo::WalRedoManager;
+use crate::ZTimelineId;
+use crate::PageServerConf;
+
+use zenith_utils::lsn::{AtomicLsn, Lsn};
+use zenith_utils::seqwait::SeqWait;
+
+mod relfile;
+
+use relfile::RelFileEntry;
+
+// Timeout when waiting or WAL receiver to catch up to an LSN given in a GetPage@LSN call.
+static TIMEOUT: Duration = Duration::from_secs(60);
+
+//
+// There is a hash table per relfile.
+//
+// For each relfile:
+//
+// Page index
+//
+//
+//
+
+pub struct InMemoryRepository {
+    conf: PageServerConf,
+    timelines: Mutex<HashMap<ZTimelineId, Arc<InMemoryTimeline>>>,
+}
+
+impl Repository for InMemoryRepository {
+
+    fn get_timeline(&self, timelineid: ZTimelineId) -> Result<Arc<InMemoryTimeline>> {
+        let timelines = self.timelines.lock().unwrap();
+
+        match timelines.get(&timelineid) {
+            Some(timeline) => Ok(timeline.clone()),
+            None => bail!("timeline not found"),
+        }
+    }
```

Comment:
The recommended style for cloning an Arc is `Arc::clone(&timeline)` rather than `timeline.clone()` to signal that we're not trying to clone the inner object.